api/client/proxy/proxy.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/net/http/httpproxy"
+)
+
+// GetProxyAddress gets the HTTP proxy address to use for a given address, if any.
+func GetProxyAddress(dialAddr string) *url.URL {
+	addrURL, err := parse(dialAddr)
+	if err != nil {
+		return nil
+	}
+	proxyFunc := httpproxy.FromEnvironment().ProxyFunc()
+	if addrURL.Scheme != "" {
+		proxyURL, err := proxyFunc(addrURL)
+		if err != nil {
+			return proxyURL
+		}
+		return nil
+	}
+
+	for _, scheme := range []string{"https", "http"} {
+		addrURL.Scheme = scheme
+		proxyURL, err := proxyFunc(addrURL)
+		if err == nil && proxyURL != nil {
+			return proxyURL
+		}
+	}
+
+	return nil
+}
+
+// parse parses a URL. If the address does not have a scheme, it will prepend "http" and try.
+func parse(addr string) (*url.URL, error) {
+	proxyurl, err := url.Parse(addr)
+	if err != nil || !strings.HasPrefix(proxyurl.Scheme, "http") {
+		proxyurl, err = url.Parse("http://" + addr)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		proxyurl.Scheme = ""
+	}
+
+	return proxyurl, nil
+}
```

Comment:
I don't think this function is needed. `url.Parse` seems to work fine without schemes as long as the URL doesn't start with a colon. And if the scheme is something strange like `mailto:`, then `"http://" + addr` is going to fail as well. I think we could just inline `url.Parse`. Is there a specific scenario this covers?

Reply:
`url.Parse` fails when given `<ip-address>:<port>` with no scheme. This fix is [similar to what httpproxy does under the hood](https://cs.opensource.google/go/x/net/+/de3da570:http/httpproxy/proxy.go;l=147).

Reply:
Ah. I tried everything except an IP address  - even "//host:1234" works :). We should return the original error though, like the proxy implementation. Any error returned from the prefixed version is not useful to callers.